src/Time/Time.hpp

```diff
@@ -0,0 +1,254 @@
+/// equality as long as they do not belong to overlapping unequal
+/// slabs.
+class Time {
+ public:
+  using rational_t = boost::rational<ssize_t>;
+
+  /// Default constructor gives an invalid Time.
+  Time() : slab_(), fraction_(0) {}
+
+  /// A time a given fraction of the way through the given slab.
+  Time(Slab slab, rational_t fraction) : slab_(slab), fraction_(fraction) {
+    range_check();
+  }
+
+  /// Move the time to a different slab.  The time must be at an end
+  /// of the current slab and the new slab must share that endpoint.
+  Time with_slab(const Slab& slab) const;
+
+  /// Approximate numerical value of the Time.
+  double value() const;
+  const Slab& slab() const { return slab_; }
+  const rational_t& fraction() const { return fraction_; }
+
+  Time& operator+=(const TimeDelta& delta);
+  Time& operator-=(const TimeDelta& delta);
+
+  bool is_at_slab_start() const { return fraction_ == 0; }
+  bool is_at_slab_end() const { return fraction_ == 1; }
+  bool is_at_slab_boundary() const {
+    return is_at_slab_start() or is_at_slab_end();
+  }
+
+  void pup(PUP::er& p);
+
+ private:
+  Slab slab_;
+  rational_t fraction_;
+
+  inline void range_check() const {
+    ASSERT(fraction_ >= 0 and fraction_ <= 1,
+           "Out of range slab fraction: " << fraction_);
+  }
+
+  friend class TimeDelta;
+};
+
+// Output
+std::ostream& operator<<(std::ostream& os, const Time& t);
+
+/// Represents an interval of time within a single slab.
+class TimeDelta {
+ public:
+  using rational_t = Time::rational_t;
+
+  /// Default constructor gives an invalid TimeDelta.
+  TimeDelta() : slab_(), fraction_(0) {}
+
+  /// An interval covering a given fraction of the slab.
+  TimeDelta(Slab slab, rational_t fraction)
+      : slab_(slab), fraction_(fraction) {}
+
+  /// Move the interval to a different slab.  The resulting interval
+  /// will in general not be the same length, but will take up the
+  /// same fraction of its slab.
+  TimeDelta with_slab(const Slab& slab) const {
+    return TimeDelta(slab, fraction());
+  }
+
+  Slab slab() const { return slab_; }
+  rational_t fraction() const { return fraction_; }
+  /// Approximate numerical length of the interval.
+  double value() const {
+    return (slab_.end_ - slab_.start_) *
+           boost::rational_cast<double>(fraction_);
+  }
+
+  /// Test if the interval is oriented towards larger time.
+  bool is_positive() const { return fraction_ > 0; }
+
+  TimeDelta& operator+=(const TimeDelta& other);
+  TimeDelta& operator-=(const TimeDelta& other);
+  TimeDelta operator+() const;
+  TimeDelta operator-() const;
+  TimeDelta& operator*=(const rational_t& mult);
+  TimeDelta& operator/=(const rational_t& div);
+
+  void pup(PUP::er& p);
+
+ private:
+  Slab slab_;
+  rational_t fraction_;
+
+  friend class Time;
+};
+
+// Output
+std::ostream& operator<<(std::ostream& os, const TimeDelta& dt);
+
+// Time <cmp> Time
+bool operator==(const Time& a, const Time& b);
```

Comment:
I think we should follow the gaming industry almost. They completely compile without exceptions but that's not an option for us. From everything I've been able to find we should `noexcept` everything we know won't throw, which is almost everything for us since we are stuck with terminate in almost all cases where we handle errors